druid-shell/src/backend/wayland/display.rs

```diff
@@ -0,0 +1,215 @@
+#![allow(clippy::single_match)]
+use super::error;
+use std::collections::BTreeMap;
+use wayland_client as wlc;
+use wayland_client::protocol::wl_registry;
+use wayland_protocols::xdg_shell::client::xdg_wm_base;
+
+#[derive(Clone)]
+pub struct GlobalEventSubscription {
+    id: u64,
+    sub: std::sync::Arc<dyn GlobalEventConsumer>,
+}
+
+impl GlobalEventSubscription {
+    fn with_id(mut self, id: u64) -> Self {
+        self.id = id;
+        self
+    }
+}
+
+impl GlobalEventConsumer for GlobalEventSubscription {
+    fn consume(
+        &self,
+        event: &wlc::GlobalEvent,
+        registry: &wlc::Attached<wl_registry::WlRegistry>,
+        ctx: &wlc::DispatchData,
+    ) {
+        self.sub.consume(event, registry, ctx)
+    }
+}
+
+impl<X> From<X> for GlobalEventSubscription
+where
+    X: Fn(&wlc::GlobalEvent, &wlc::Attached<wl_registry::WlRegistry>, &wlc::DispatchData)
+        + GlobalEventConsumer
+        + 'static,
+{
+    fn from(closure: X) -> Self {
+        Self {
+            id: 0,
+            sub: std::sync::Arc::new(closure),
+        }
+    }
+}
+
+impl<X> GlobalEventConsumer for X
+where
+    X: Fn(&wlc::GlobalEvent, &wlc::Attached<wl_registry::WlRegistry>, &wlc::DispatchData) + 'static,
+{
+    fn consume(
+        &self,
+        event: &wlc::GlobalEvent,
+        registry: &wlc::Attached<wl_registry::WlRegistry>,
+        ctx: &wlc::DispatchData,
+    ) {
+        self(event, registry, ctx)
+    }
+}
+
+pub trait GlobalEventDispatch {
+    fn subscribe(&self, sub: impl Into<GlobalEventSubscription>) -> GlobalEventSubscription;
+    fn release(&self, s: &GlobalEventSubscription);
+}
+
+pub trait GlobalEventConsumer {
+    fn consume(
+        &self,
+        event: &wlc::GlobalEvent,
+        registry: &wlc::Attached<wl_registry::WlRegistry>,
+        ctx: &wlc::DispatchData,
+    );
+}
+
```

Comment:
I don't understand the purpose of this trait, and `GlobalEventSubscription` .

why can't `Dispatcher` just `BTreeMap<u64, Arc<dyn Fn(.......)>>` (using a type alias for `Arc<dyn ....>` for making it look readable)

Reply:
original intention was to allow you to not rely on the particular implemention when processing events. I can probably remove it and collapse everything.